plugins/experimental/remap_stats/remap_stats.cc

```diff
@@ -160,81 +140,80 @@ handle_post_remap(TSCont cont, TSEvent event ATS_UNUSED, void *edata)
   return 0;
 }
 
-#define CREATE_STAT_NAME(s, h, b) snprintf(s, MAX_STAT_LENGTH, "plugin.%s.%s.%s", PLUGIN_NAME, h, b)
+static std::string
+create_stat_name(const std::string &h, const std::string &b)
+{
+  std::string res(std::string("plugin.") + PLUGIN_NAME + '.' + h + '.' + b);
+  return res.substr(0, MAX_STAT_LENGTH);
+}
 
 static int
 handle_txn_close(TSCont cont, TSEvent event ATS_UNUSED, void *edata)
 {
-  TSHttpTxn txn = (TSHttpTxn)edata;
+  TSHttpTxn txn = static_cast<TSHttpTxn>(edata);
   config_t *config;
   void *txnd;
   int status_code = 0;
   TSMBuffer buf;
   TSMLoc hdr_loc;
   uint64_t out_bytes, in_bytes;
-  char *remap, *hostname;
-  char *unknown = "unknown";
-  char stat_name[MAX_STAT_LENGTH];
+  std::string remap, hostname;
```

Comment:
These three should be `std::string_view`. Note the original code presumes them to be C strings.

Reply:
Ok, would it make sense for `stat_name` to be `std::string_view` too? Or just `remap`, `hostname` and `unknown`?

Reply:
The primary reason to use `string_view` instead of `string` is to avoid unnecessary memory allocation. I would be tempted to keep the original `char stat_name[MAX_STAT_LENGTH];` for that reason. Or if you want to use `std::string`, then `reserve` that amount of space to have only 1 allocation. You can pass a `string_view` to something that expects `const char*`, you just need to be careful about whether it's expected to be a C string. In this particular case, the original logic clearly assumed that.

You could also look at `ts::LocalBufferWriter` which was written for this situation. E.g.
```
ts::LocalBufferWriter<MAX_STAT_LENGTH> stat_w;
// ...
stat_w.reset().reduce(1); // clear any current data, reserve 1 byte for later.
stat_w.print("plugin.{}.{}.{}", PLUGIN_NAME, remap, "in_bytes");
stat_w.extend(1).write('\0'); // get back the byte and null terminate the string.
```
The `reduce` / `extend` calls make sure the name is a null terminated C string even if the
output is larger than `MAX_STAT_LENGTH`. `LocalBufferWriter` stack allocates its internal buffer and clips output to that size.

Reply:
`ts::LocalBufferWriter` seems to be perfect for this, should I pass a `const ts::LocalBufferWriter &` to `stat_add` though? Or a `const char *` would be better? After all I'll always use it like a C string in that method. If I call `stat_add` with `const ts::LocalBufferWriter &` should also use it as a key for the internal hash?

I'm leaning toward calling `stat_add` with `const char *` but I want to know your opinion too.

Reply:
You should definitely not call it with `ts::LocalBufferWriter`. Generally you would use the `view` method to pass a `string_view` after writing to the buffer. In this case, since it's dealing with a predefined low level C interface, I would go ahead and pass in a `const char*`, via `stat_w.data()`.

Reply:
Ah, didn't see the key for a hash question. The Buffer Writer stuff is useful only for writing to buffers, it's not suitable for persistence. The normal use for that is to use `string_view` as the key and get use the `view` method of the Buffer Writer for it. In these cases you need to be careful of the buffer lifetime. If you need it to persist (so as to serve as a key, for instance), you should use `ts::FixedBufferWriter` which writes on an externally defined buffer. You could also use `ts::bwprint` which does the busy work of doing all this to a `std::string`. Then you can use that `string` as the key. E.g.
```
std::string stat_name;
ts::bwprint(stat_name, "plugin.{}.{}.{}", ...);
```
That takes care of resizing the `string` as needed and you get the null termination without explicit work. It does potentially require an allocation, but that's probably OK here since this is startup and not run time code.

I really need to improve the documentation on this, although there is some - https://docs.trafficserver.apache.org/en/latest/developer-guide/internal-libraries/buffer-writer.en.html